apps/explorer/src/utils/thumbnails.ts

```diff
@@ -0,0 +1,185 @@
+const thumbnailBlobCache: Record<string, Blob | undefined> = {}
+
+export const fetchVideoBlob = async (videoUrl: string): Promise<Blob> => {
+  if (!videoBlobCache[videoUrl]) {
+    videoBlobCache[videoUrl] = fetch(videoUrl)
+      .then((response) => {
+        if (!response.ok) {
+          throw new Error(`Failed to fetch video: ${response.statusText}`)
+        }
+        return response.blob()
+      })
+      .catch((error) => {
+        delete videoBlobCache[videoUrl]
+        throw error
+      })
+  }
+
+  return videoBlobCache[videoUrl]!
+}
+
+export const createThumbnailFromVideoBlob = (blob: Blob): Promise<Blob> =>
+  new Promise((resolve, reject) => {
+    const video = document.createElement('video')
+    const url = URL.createObjectURL(blob)
+    video.src = url
+    video.muted = true // Required for autoplay in some browsers
+    video.crossOrigin = 'Anonymous'
+    video.playsInline = true // Required for iOS to allow autoplay
+
+    const handleLoadedData = () => {
+      if (video.videoWidth === 0 || video.videoHeight === 0) {
+        reject(new Error('Invalid video dimensions'))
+        return
+      }
+
+      const canvas = document.createElement('canvas')
+      const scale = 0.5 // Reduce resolution by 50%
+      canvas.width = video.videoWidth * scale
+      canvas.height = video.videoHeight * scale
+      const context = canvas.getContext('2d')
+
+      if (!context) {
+        reject(new Error('Failed to get canvas context'))
+        return
+      }
+
+      const drawFrame = () => {
+        setTimeout(() => {
+          context.drawImage(video, 0, 0, canvas.width, canvas.height)
+          canvas.toBlob((thumbnailBlob) => {
+            if (thumbnailBlob) {
+              resolve(thumbnailBlob)
+            } else {
+              reject(new Error('Failed to create thumbnail blob'))
+            }
+            video.remove()
+            URL.revokeObjectURL(url)
+          })
+        }, 100) // Adding a small delay to ensure the frame is ready (mobile)
+      }
+
+      requestAnimationFrame(() => {
+        requestAnimationFrame(drawFrame)
+      })
+    }
+
+    video.addEventListener('loadeddata', () => {
+      if (video.readyState >= 2) {
+        handleLoadedData()
+      } else {
+        video.currentTime = 0 // Seek near the beginning
+        video.addEventListener('seeked', handleLoadedData, { once: true })
+      }
+    })
+
+    video.addEventListener('error', (e) => {
+      reject(new Error(`Failed to load video: ${e.message}`))
+    })
+
+    video.play() // Force video play in case autoplay is blocked
+  })
+
+export const isValidThumbnail = (blob: Blob): boolean => blob.size > 17000 // Basic test. Approx size of blank thumbnail.
+
+export const getOrCreateThumbnail = async (videoUrl: string): Promise<Blob> => {
+  if (thumbnailBlobCache[videoUrl]) {
+    const blob = thumbnailBlobCache[videoUrl]!
+    if (isValidThumbnail(blob)) {
+      return blob
+    }
+  }
```

Comment:
Why the `!` at the end? ChatGPT? 😆 

Wouldn't it be simpler to write:

```ts
  const blob = thumbnailBlobCache[videoUrl]
  if (blob) return blob
```

Reply:
No, it's an artefact of my many iterations I believe. I had to do this at one point. Will check again!